src/libraries/Microsoft.PowerFx.Interpreter/Functions/CollectFunction.cs

```diff
@@ -0,0 +1,362 @@
+using Microsoft.PowerFx.Core.Types;
+using Microsoft.PowerFx.Core.Utils;
+using Microsoft.PowerFx.Syntax;
+using Microsoft.PowerFx.Types;
+using static Microsoft.PowerFx.Core.Localization.TexlStrings;
+
+namespace Microsoft.PowerFx.Interpreter
+{
+    // The CollectFunction class was inspired (copied) by Set function.
+    // Implementation of a Set function which just chains to 
+    // RecalcEngine.UpdateVariable().
+    // Set has no return value. 
+    // Whereas PowerApps' Set() will implicitly define arg0,
+    //  this Set() requires arg0 was already defined and has a type.
+    //
+    // Called as:
+    //   Set(var,newValue)
+    internal class CollectFunction : BuiltinFunction, IAsyncTexlFunction
+    {
+        public override bool SuppressIntellisenseForComponent => true;
+
+        public override bool ManipulatesCollections => true;
+
+        public override bool ModifiesValues => true;
+
+        public override bool AffectsCollectionSchemas => true;
+
+        public override bool IsSelfContained => true;
+
+        public override bool CanSuggestThisItem => true;
+
+        public override bool RequiresDataSourceScope => true;
+
+        protected virtual bool IsScalar => false;
+
+        public override bool SupportsParamCoercion => false;
+
+        public override bool DisableForCommanding => true;
+
+        public override bool ArgMatchesDatasourceType(int argNum)
+        {
+            return argNum >= 1;
+        }
+
+        public CollectFunction()
+        : base(
+              DPath.Root,
+              "Collect",
+              "Collect",
+              TexlStrings.AboutSet,
+              FunctionCategories.Behavior,
+              DType.EmptyTable,
+              0, // no lambdas
+              2,
+              int.MaxValue)
+        {
+        }
+
+        public override IEnumerable<TexlStrings.StringGetter[]> GetSignatures()
+        {
+            // PR REVIEWERS: These are wrong signature texts.
+            yield return new[] { TexlStrings.WithArg1, TexlStrings.WithArg2 };
+            yield return new[] { TexlStrings.WithArg1, TexlStrings.WithArg2, TexlStrings.WithArg2 };
+            yield return new[] { TexlStrings.WithArg1, TexlStrings.WithArg2, TexlStrings.WithArg2, TexlStrings.WithArg2 };
+        }
+
+        public override IEnumerable<TexlStrings.StringGetter[]> GetSignatures(int arity)
+        {
+            if (arity > 2)
+            {
+                return GetGenericSignatures(arity, TexlStrings.WithArg1, TexlStrings.WithArg2);
+            }
+
+            return base.GetSignatures(arity);
+        }
+
+        public virtual DType GetCollectedType(DType argType)
+        {
+            Contracts.Assert(argType.IsValid);
+
+            return argType;
+        }
+
+        // Attempt to get the unified schema of the items being collected by an invocation.
+        public bool TryGetUnifiedCollectedType(TexlNode[] args, DType[] argTypes, IErrorContainer errors, out DType collectedType)
```

Comment:
We don't want to do implicit collection schema definition for this use, likely all this type checking is irrelevant.

Reply:
Thanks for pointing that. By removing this method, the following also becomes irrelevant and can also be delete (from CheckInvocation method).
```
            fValid &= TryGetUnifiedCollectedType(args, argTypes, errors, out DType collectedType);
            Contracts.Assert(collectedType.IsTable);

            // The item type must be compatible with the collection schema.
            var fError = false;
            returnType = DType.Union(ref fError, collectionType, collectedType, useLegacyDateTimeAccepts: true);
            if (fError)
            {
                fValid = false;
                if (!SetErrorForMismatchedColumns(collectionType, collectedType, args[1], errors))
                {
                    errors.EnsureError(DocumentErrorSeverity.Severe, args[0], TexlStrings.ErrNeedValidVariableName_Arg);
                }
            }
```

Make sense?